lib/paquito/codec_factory.rb

```diff
@@ -5,15 +5,24 @@
 
 module Paquito
   class CodecFactory
-    def self.build(types, freeze: false, serializable_type: false)
-      factory =
-        if types.empty? && !serializable_type
-          MessagePack
-        else
-          MessagePack::Factory.new
+    def self.build(types, freeze: false, serializable_type: false, pool: nil)
+      factory = if types.empty? && !serializable_type
+        MessagePack::DefaultFactory
+      else
+        MessagePack::Factory.new
+      end
+
+      if pool
+        if serializable_type || types.any? { |t| Types.recursive?(t) }
+          pool *= 2
```

Comment:
So the pool size being two assumes no more than depth 1 recursion, which is reasonable but maybe should be mentioned somewhere?